core/chains/evm/client/chain_client.go

```diff
@@ -272,10 +279,57 @@ func (c *chainClient) LatestFinalizedBlock(ctx context.Context) (*evmtypes.Head,
 }
 
 func (c *chainClient) CheckTxValidity(ctx context.Context, from common.Address, to common.Address, data []byte) *SendError {
-	msg := ethereum.CallMsg{
-		From: from,
-		To:   &to,
-		Data: data,
+	reqs := []TxSimulationRequest{
+		{
+			From: from,
+			To:   &to,
+			Data: data,
+		},
+	}
+	err := c.BatchCheckTxValidity(ctx, reqs)
+	if err != nil {
+		return NewSendError(err)
+	}
+	return reqs[0].Error
+}
+
+func (c *chainClient) BatchCheckTxValidity(ctx context.Context, reqs []TxSimulationRequest) error {
+	rpcBatchCalls := make([]rpc.BatchElem, len(reqs))
+	for i, req := range reqs {
+		toAddress := common.Address{}
+		if req.To != nil {
+			toAddress = *req.To
+		}
+		rpcBatchCalls[i] = rpc.BatchElem{
+			Method: "eth_estimateGas",
+			Args: []any{
+				map[string]interface{}{
+					"from": req.From,
+					"to":   toAddress,
+					"data": hexutil.Bytes(req.Data),
+				},
+				"pending",
+			},
+			Result: new(hexutil.Big),
+		}
+	}
+	err := c.BatchCallContext(ctx, rpcBatchCalls)
+	if err != nil {
+		return err
+	}
+
+	for _, elem := range rpcBatchCalls {
+		params := elem.Args[0].(map[string]interface{})
+		for i := 0; i < len(reqs); i++ {
+			req := &reqs[i]
+			// Match the request to rpc response to set the error in the proper object
+			if params["from"] == req.From &&
```

Comment:
What is this check for? IIUC, you're checking req against params, which comes from rpcBatchCalls.Args and it should it always be true, as it doesn't get mutated, only rpcBatchCalls.Result does.

Reply:
Originally I was using it as a sanity check to make sure I was setting the error in the right req but you're right. The order doesn't change so this nested looping doesn't serve a purpose. I'll get this changed.

Reply:
Updated this in the following commit